site/content/extend/plugins/signing-plugins.md

```diff
@@ -0,0 +1,151 @@
+Mattermost server will verify plugin signatures downloaded from Mattermost Marketplace. Mattermost public key is shipped with the server. If needed one can download and verify plugins from custom marketplace as well. To do this, custom public key has to be added on the server. For this purpose, start the Mattermost server and run the following command:
+```console
+> mattermost plugin add key my-pub-key
+```
+
+Multiple public keys can be added to the Mattermost server:
+```console
+> mattermost plugin add key my-pk-file1 my-pk-file2
+```
+To list the names of all public keys installed on your Mattermost server, use:
+```console
+> mattermost plugin keys
+```
+To delete public key(s) from your Mattermost server, use:
+```console
+> mattermost plugin delete key my-pk-file1 my-pk-file2
+```
+
+### Storing Plugin Signatures
+When a plugin is installed, Mattermost server will store plugin signatures in its internal filestore (not to be confused with the public keys). This is to allow the server to verify signatures during the plugin-synchronization process that happens during server startup. To avoid name collisions and overwriting existing files, Mattermost will store plugin signatures alongside plugin bundles using the naming convention mentioned below. The benefit of storing signature files independently is to avoid double-tar, where the tar file contains the bundle tar along with its signature. This would have required Mattermost to migrate existing plugin bundles into double-tar format. Which would have resulted in increased complexity and made upgrade/downgrade scenarios a lot harder to deal with.
+```console
+# Filestore plugin structure
+/plugins/<plugin_id>.tar.gz            # plugin bundle tar
+/plugins/<plugin_id>.<counter-1>.sig   # multiple plugin signatures
+/plugins/<plugin_id>.<counter-2>.sig     
+
+# Examples
+/plugins/com.mattermost.demo.tar.gz    # plugin bundle tar
+/plugins/com.mattermost.demo.1.sig     # multiple plugin signatures
+/plugins/com.mattermost.demo.2.sig
+/plugins/com.mattermost.demo.3.sig
+/plugins/com.github.matterpoll.tar.gz
+/plugins/com.github.matterpoll.1.sig
+/plugins/com.github.matterpoll.2.sig
+/plugins/com.github.matterpoll.3.sig
+
+```
+
+### Mattermost Plugin Assets Naming Convention
+In order to keep plugin signatures close to their bundle tar, Mattermost will generate plugin signatures and upload them directly to GitHub releases using the following naming convention. This is purely a convention which makes it easier for a human to parse signature and public key pairs at a quick glance.
+```console
+# Mattermost plugin naming convention
+<plugin_id>-<plugin_version>.tar.gz
+<plugin_id>-<plugin_version>-<counter>-<public_key_fingerprint_hash>.sig
+
+
+# Mattermost demo plugin example
+com.mattermost.demo-plugin-0.2.0.tar.gz
+com.mattermost.demo-plugin-0.2.0-1-1E11.sig
+com.mattermost.demo-plugin-0.2.0-2-6E1D.sig
+com.mattermost.demo-plugin-0.2.0-3-3MN6.sig
+```
+
+### Configuration
```

Comment:
I'd suggest pulling this up earlier in the document to draw emphasis to the high-level details before diving in too deep.